Adding FNO and DeepONets to Neuromancer

Parv621 opened the pull request:
This PR updates **#263**.

It introduces **DeepONets** and **Fourier Neural Operators (FNOs)** into **Neuromancer**, with integrations to the external libraries **[neuralop](https://github.com/neuraloperator/neuraloperator)** and **[DeepXDE](https://deepxde.readthedocs.io/en/latest/)**

Integration is provided via lightweight wrappers that preserve Neuromancer’s abstractions while maintaining compatibility with upstream libraries.

---

## Summary of Changes

### Operator Integrations

1. **FNO Wrapper**
   `src/neuromancer/modules/operators.py`
   A thin wrapper enabling direct use of `neuralop` models inside Neuromancer. The wrapper attaches the required `_metadata` for robust checkpointing across PyTorch versions. The design generalizes to any operator provided by `neuralop`.

2. **Native DeepONet (Cartesian Product)**
   `src/neuromancer/modules/operators.py`
   Native implementation of DeepONet with Cartesian-product evaluation, consistent with standard operator-learning formulations.

3. **DeepXDE Wrapper**
   `src/neuromancer/modules/operators.py`
   Wrapper for DeepONet-style models implemented in DeepXDE. Supports both
   `dde.nn.DeepONetCartesianProd` and `dde.nn.DeepONet`.

4. **Custom Operator Losses**
   `src/neuromancer/modules/operators.py`
   Implementation of `LpLoss` and `H1Loss`, suitable for grid-based operator learning.

---

### Training and Validation Infrastructure

5. **Validation Dataset Support During Training**
   Changes in `trainer.py`, `problem.py`, and `callbacks.py` to support validation datasets during training (inspired by #168).
   A new `LossHistoryCallback` is added to `callbacks.py` to track and plot training and validation loss per epoch.

---

## Tests Added

The following unit tests were added to validate new functionality:

- **`test_trainer_optional_dev_data`** (`test_trainer.py`)  
  Verifies that the training loop runs correctly both with and without a validation (dev) dataset.

- **`test_problem_predict`** (`test_problem.py`)  
  Tests the `Problem.predict` method, ensuring gradients are disabled during inference and that input/output keys are handled correctly.

- **`test_strip_state` and `test_strip_load`** (`test_operators.py`)  
  Validate the `_StripMetadataMixin` used by the `neuralop` wrapper to ensure metadata is correctly stripped during save and load operations.

- Tests were added for `DeepXDEWrapper` and `DeepONetCartesianProd` in `operators.py`, (`test_operators.py`)  .

---

## Implemented Notebooks

### DeepONet Examples

1. **Antiderivative (Aligned GRF Dataset)**
   `examples/neural_operators/Part_1_DeepONet_antiderivative_aligned.ipynb`
   Inspiration: DeepXDE aligned antiderivative demo and Lu et al. (2021).

2. **Antiderivative (Unaligned GRF Dataset)**
   `examples/neural_operators/Part_2_DeepONet_antiderivative_unaligned.ipynb`
   Inspiration: DeepXDE unaligned antiderivative demo.

3. **1D Advection equation (Aligned initial conditions from GRF fields)**
   `examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb`
   Inspiration: DeepXDE aligned 1D Advection demo. 

---

### FNO Examples

1. **1D Allen–Cahn Equation (from-scratch FNO)**
   `examples/neural_operators/Part_4_FNO_1DAllenCahn.ipynb`
   Inspiration: ETH “Operator Learning – FNO” tutorial and the original FNO paper.

2. **Darcy Flow (FNO, Neuromancer pipeline)**
   `examples/neural_operators/Part_5_FNO_DarcyFlow.ipynb`
   Inspiration: `neuralop` Darcy Flow example and NVIDIA’s FNO Darcy guide.

3. **Diffusion Equation (FNO, Neuromancer pipeline)**
   `examples/neural_operators/Part_6_FNO_DiffusionEquation.ipynb`
   Inspiration: `examples/PDEs/Part_1_PINN_DiffusionEquation.ipynb`.

4. **Shallow Water equations using Spherical FNO (SFNO, Neuromancer pipeline)**
   `examples/neural_operators/Part_7_SFNO_ShallowWater.ipynb`
   Inspiration: `neuralop` SFNO on the spherical Shallow Water equations.

---

## Notebooks I could not implement (DeepONet)

The following notebooks are not included as they do not achieve accurate solutions using standard MLP DeepONets:

* **Cartesian-Product DeepONet – Darcy Flow**
* **Cartesian-Product DeepONet – Diffusion Equation**
* **Vanilla DeepONet – Darcy Flow**
* **Vanilla DeepONet – Diffusion Equation**

These are moved to another SOLARIS branch (`feature/DeepONet_notebooks_debug`); we could revisit them later.

Some papers try to solve DarcyFlow using NOs, might be useful for debugging. [NVIDIA example on Darcy Flow using DeepONet](https://docs.nvidia.com/physicsnemo/25.08/physicsnemo-sym/user_guide/neural_operators/deeponet.html#problem-3-darcy-flow-data-informed) uses a convnet-based model [pix2pix](https://docs.nvidia.com/physicsnemo/25.08/physicsnemo-sym/user_guide/theory/architectures.html#pix2pix-net) in the branch net. Another paper (https://www.mdpi.com/2227-9717/13/9/2754) solves DarcyFlow using various methods

Parv621 committed:
First code for 1D FNO added

Parv621 committed:
FNO testing on PINNs notebook - data ready

Parv621 committed:
Data symbolic input working, Training part left

Parv621 committed:
Moved to NO folder, working code (on diffusion). Need to fix training, loss defintions etc

Parv621 committed:
Updated descriptions, working version for L2 loss, H1 still not working

Parv621 committed:
Darcy flow example from neuralop, figured data loaders

Parv621 committed:
Working Darcy flow example with FNO, only with L2 loss

Parv621 committed:
Added wrapper for neuralop.models.FNO to remove metadata

Parv621 committed:
Wroking notebooks with L2 loss. Added init call for operators

Parv621 committed:
Dimenisionality fixes for y_grid/y_fno. Actually improves predictions

Parv621 committed:
Added H1loss nn.modules to operators, make it work in DarcyFlow example

Parv621 committed:
Use Lploss, and H1loss as objects instead of nn.module

Parv621 committed:
Fix docstrings, use new L2 loss and H1 loss in notebooks

Parv621 committed:
Add 16 x 16 test, cleanup notebook

Parv621 committed:
Cleanup operators.py, Add docstrings to loss objects

Parv621 committed:
Final cleanup before merge request

Parv621 committed:
Remove FNOWithoutMeta, update docstring

Parv621 committed:
Added 3 different approaches for H1loss in Darcy Flow notebook

Parv621 committed:
Integrate feature/DeepONet changes into feature/FNO

Squashed commit of the following:

commit f39161f6a84971ee2ee1df3cb5ef3f838e63ede3
Author: parv621@gmail.com <parv621@gmail.com>
Date:   Fri Nov 28 18:31:55 2025 -0500

    Moved DeepXDEDataWrapper to operator.spy, final DeepXDE notebook cleanup

commit 022c59019510665acada226b439269df0b086e0a
Author: parv621@gmail.com <parv621@gmail.com>
Date:   Fri Nov 28 18:11:34 2025 -0500

    Working version of anti-derivative with DeepXDE DeepONetCartesianProd. Create correct data pipeline

commit bc5b4c5cabda5ff8359b12d0ce27c71e9a8d866a
Author: parv621@gmail.com <parv621@gmail.com>
Date:   Fri Nov 28 11:32:11 2025 -0500

    Final notebook for antiderivative example

commit a83ff8e72f41aef5cdd10dd775f76db42df01a98
Author: parv621@gmail.com <parv621@gmail.com>
Date:   Fri Nov 28 11:14:37 2025 -0500

    use Einsum instead of matmul for more general dim_x cases

commit c7ae9fa55441544d08f3ad28174cf8afa6a6ecc8
Author: parv621@gmail.com <parv621@gmail.com>
Date:   Thu Nov 27 19:27:42 2025 -0500

    Antiderivative notebook working version. Minimal changes in operators.py

commit d31edba8813a356aa1777dd61aeab60821da1f6b
Merge: 1f97d43 d93bf89
Author: parv621@gmail.com <parv621@gmail.com>
Date:   Thu Nov 27 11:35:20 2025 -0500

    Merge branch 'master' into feature/DeepONet

commit 1f97d43372d654c67bd4a6826c1e754004f344b6
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Wed Aug 28 16:46:56 2024 -0700

    remove data_dir and cut some old unused code

commit 4e01c8fb7944724376418c79408da6fe742e7aa3
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Fri Aug 2 17:01:56 2024 -0700

    Get model to work with generated data

commit 2f02dcfc7e1e8ec206be7c7594affe81acdffaec
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Mon Jul 29 09:38:11 2024 -0700

    Add attempt at generating 1d data and antiderivative ground truth

commit b3206d4b5de91071f22b32ad02d00f094cc8384a
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Mon Jul 29 09:37:47 2024 -0700

    add unfinished comments about shapes of DeepONet

commit 0de774faaf6b8cf4709ff039ef82673c8dd25f7c
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Tue Jul 2 15:16:52 2024 -0700

    Remove extra code for getting data generation working from operators.py

commit ed34ee23bd0c8cd7a2170d0b28f28fb5c92447e8
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Tue Jul 2 15:03:50 2024 -0700

    Push changes from trainer

commit 032ae78a3e9787f13b43df66d9c6b80f56ee1a11
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Tue Jul 2 15:01:57 2024 -0700

    Test changes on problem to see if autoformat issue

commit 17f62bd6e50fd0bb49165217a88352a11a18f4a7
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Tue Jul 2 15:00:06 2024 -0700

    Attempt to add problem and trainer back in without format on save

commit 57beff6b94fc46f5e9b677d12c3d497685af48aa
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Tue Jul 2 14:56:22 2024 -0700

    Delete local version of my specific changes with too much formatting

commit 1c08c963aaa535aedaab2cefd16cccaa4fed42b8
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Tue Jul 2 14:55:15 2024 -0700

    Move changes from problem back

commit e84e2b123d126242fd78bf79e8f2edcd82c711ce
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Tue Jul 2 14:54:11 2024 -0700

    Move changes from trainer back

commit ded1a01bc5610ee991a9b847e931beaaf5219762
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Tue Jul 2 14:51:53 2024 -0700

    Add trainer and problem from develop, add separate copy i'm working on

commit ab1f644a2d907c18e94adfd68674c289447aa2cb
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Sat Jun 29 12:00:50 2024 -0700

    Attempt to get data generation working

commit 8a3c4692e0d7b93e7bf34070272eb0bc39b130c3
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Fri Jun 28 17:12:23 2024 -0700

    Add latest progress working on data generation

commit 3f9fb6cb9c181a68957178ff757d58ef7fc8e311
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Fri Jun 28 16:01:21 2024 -0700

    Remove dataset download code from notebook

commit 3b66c3c17f14adbe725110ce97dd6297b1074f43
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Fri Jun 28 15:58:54 2024 -0700

    Finish initial Class documentation for DeepONet

commit be5f2c091fa736250e0ff9250fd3785f7506b898
Author: Colby Ham <colby.ham@pnnl.gov>
Date:   Fri Jun 28 13:38:21 2024 -0700

    Add current state of example from feature/deeponet_examples after messy rebase

Parv621 committed:
DeepONet Diffusion notebook - Same as PINNs

Parv621 committed:
Updated the nm DeepONet implementation to be batch first (PyTorch convention)

Parv621 committed:
Updated antiderivative notebooks (in-notebook DeepONet, updated docstrings)

Parv621 committed:
Moved DeepONet code from dynamics.operators to modules.operators

Parv621 committed:
Moved notebooks to neural_operators folder. Rename

Parv621 committed:
Create .py script for the DeepOnet antiderivative notebooks in examples/neural_operators

Parv621 committed:
Scripts for FNO notebooks

Parv621 committed:
Markddowns for problem context - FNO

Parv621 committed:
Added DeepONetBatchWrapper for dde.nn.DeepONet that works

Parv621 committed:
Diffusion example with DeepONet that trains, but produces garbage results

drgona reviewed (commented):
we can add the `LossHistoryCallback` class to callbacks.py

Since users will start from the first example, it would be good if the Part 1 example has extensive method markdown, similar to Part 3.

Parv621 committed:
Added LossHistoryCallback in callback.py, plots dev/train losses on a graph

Parv621 committed:
Added DeepXDE to aligned notebook, combined notebooks. Add description and references in markdowns

Parv621 commented:
> we can add the `LossHistoryCallback` class to callbacks.py
> 
> Since users will start from the first example, it would be good if the Part 1 example has extensive method markdown, similar to Part 3.

Added in callback in bf02c6119214968d731275a52e1dd1c166962126, updated Part 1 in 002d89c1b5b237302dbeb3ac55da44d2e31d16e3

Parv621 committed:
Update .py for aligned example

Parv621 committed:
Start working on unaligned example

Parv621 committed:
Fully working unaligned example

Parv621 committed:
new unified DeepXDE wrappers, use them in notebooks

Parv621 committed:
Create CartesianProd DeepONet Darcy Flow notebook - the standard DeepONet doesnt train well

Parv621 committed:
Current struggles with DeepONets

Parv621 committed:
operators.py cleanup

Parv621 committed:
1D FNO from scratch - Allen Cahn example with documentation

Parv621 committed:
Complete FNO from scratch Allen Cahn notebook

Parv621 committed:
Rename files for correct order in examples folder

Parv621 committed:
Further file renaming

Parv621 committed:
Removed FNO1D implementation from blocks. Neuralop will be used until we need our own implementation

Parv621 committed:
notebook cleanup

Parv621 commented on pyproject.toml:
```diff
@@ -43,14 +43,14 @@ dependencies = [
     "pydot==1.4.2",
     "pyts",
-    "torch",
+    "torch>=1.8",
```
For ifft in PyTorch needed for FNOs

Parv621 replied:
Should I add dependencies for neuralop and deepxde to the environment? Or would you like to keep it flexible?

drgona replied:
It depends how many and what dependencies there are. If it is a lot, we can add the necessary dependencies for neuralop and deepxde to examples dependencies in our toml file.

https://github.com/pnnl/neuromancer/blob/master/pyproject.toml#L69

Parv621 replied:
So, the `deepxde` library can be added like that, as we use it only in examples.
The `neuraloperator` library is being used here https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/src/neuromancer/modules/operators.py#L6, so it will either need to be added as a main dependency. Or we could do something like this in pyproject.toml file, loading this only for operators.

```toml
examples = [
    "deepxde",
    "imageio",
    "neuraloperator",
]
[project.optional-dependencies]
operators = [
  "neuraloperator",
]
```

Let me know what you prefer!

Parv621 commented on src/neuromancer/modules/blocks.py:
Formatting changes only (linter)

drgona commented:
lets run all the unit tests on your on this PR to make sure we are not braking the core

drgona commented:
Well done with the examples!
Few minor comments on notebooks below.

1. Lets make sure the examples are ordered withour overlapping Part numbers. Lets consolidate examples to make them more compact. 
2. Each new example should add markdown to new features or architectures it introduces compared to previous examples.
3. Can we have antiderivative example to introduce simple FNO as well?

**Part_1_DeepONet_antiderivative_aligned.ipynb**
markdown cell with Operator evaluation: lets specify `b` for grid point evaluations, similarly we define `m` for sensor points
lets explain what is the attribute  "Glorot normal" in DeepXDE DeepOnet

**Part_2_DeepONet_antiderivative_unaligned.ipynb**
it would be nice to have base DeepONet class in neuromancer and compare with DeepXDE in this example similarly as in part 1. Nice to have if easy to implement, but not critical.

**Part_5_FNO_DarcyFlow.ipynb**
If we don't introduce FNO on antiderivative, we can use this example as main intro to FNO and make it part 3 and potentially merge with **Part_3_DeepONet_DarcyFlow.ipynb** to compare FNO vs DeepOnet

**Part_3_DeepONet_DarcyFlow.ipynb**
this notebook introduces for the first time `DeepONetBatchWrapper ` but it is not explained, what is the difference between the `DeepXDEWrapper`?

It is not apparent from the notebooks what is the main difference between:
**Part_3_DeepONetCartesianProd_DarcyFlow.ipynb**
**Part_3_DeepONet_DarcyFlow.ipynb**
Maybe we can merge these two into one notebook and compare cartesiaon prod with non-catesian DeepOnet and focus on explaining the main differences and their implications.
Currently both notebooks have large cells of code without markdown.

Similarly:
**Part_6_FNO_DiffusionEquation.ipynb**
**Part_3_DeepONet_DiffusionEquation.ipynb**
**Part_3_DeepONetCartesianProd_DiffusionEquation.ipynb**
These can be merged into single exmaple that compares FNO vs DeepONet on diffusion example and focus on explaining main differences.

**Part_3_DeepONetCartesianProd_DiffusionEquation.ipynb** this title does not match the content which shows FNO. It looks like deprecated file to be deleted.

**Part_4_FNO_1DAllenCahn.ipynb**
The role of convolution layers should be explained.

drgona commented on src/neuromancer/callbacks.py:
lets make sure this dependency is included in our toml file:
from pathlib import Path

Parv621 replied:
pathlib is a standard library module. No dependency entry is required.

drgona commented on src/neuromancer/problem.py:
since you introduce change to the core abstraction, we need to run all the tests to make sure we are not breaking the core

Parv621 replied:
<img width="1946" height="294" alt="image" src="https://github.com/user-attachments/assets/740a8c8a-6bc8-4f17-8a94-6aa5f6a51ca8" />

drgona commented on src/neuromancer/trainer.py:
similarly here, lets run all the tests to make sure these changes to the core abstraction are backward compatible

Parv621 replied:
> similarly here, lets run all the tests to make sure these changes to the core abstraction are backward compatible

this passed with a few warnings

drgona commented:
> Well done with the examples! Few minor comments on notebooks below.
> 
> 1. Lets make sure the examples are ordered withour overlapping Part numbers. Lets consolidate examples to make them more compact.
> 2. Each new example should add markdown to new features or architectures it introduces compared to previous examples.
> 3. Can we have antiderivative example to introduce simple FNO as well?
> 
> **Part_1_DeepONet_antiderivative_aligned.ipynb** markdown cell with Operator evaluation: lets specify `b` for grid point evaluations, similarly we define `m` for sensor points lets explain what is the attribute "Glorot normal" in DeepXDE DeepOnet
> 
> **Part_2_DeepONet_antiderivative_unaligned.ipynb** it would be nice to have base DeepONet class in neuromancer and compare with DeepXDE in this example similarly as in part 1. Nice to have if easy to implement, but not critical.
> 
> **Part_5_FNO_DarcyFlow.ipynb** If we don't introduce FNO on antiderivative, we can use this example as main intro to FNO and make it part 3 and potentially merge with **Part_3_DeepONet_DarcyFlow.ipynb** to compare FNO vs DeepOnet
> 
> **Part_3_DeepONet_DarcyFlow.ipynb** this notebook introduces for the first time `DeepONetBatchWrapper ` but it is not explained, what is the difference between the `DeepXDEWrapper`?
> 
> It is not apparent from the notebooks what is the main difference between: **Part_3_DeepONetCartesianProd_DarcyFlow.ipynb** **Part_3_DeepONet_DarcyFlow.ipynb** Maybe we can merge these two into one notebook and compare cartesiaon prod with non-catesian DeepOnet and focus on explaining the main differences and their implications. Currently both notebooks have large cells of code without markdown.
> 
> Similarly: **Part_6_FNO_DiffusionEquation.ipynb** **Part_3_DeepONet_DiffusionEquation.ipynb** **Part_3_DeepONetCartesianProd_DiffusionEquation.ipynb** These can be merged into single exmaple that compares FNO vs DeepONet on diffusion example and focus on explaining main differences.
> 
> **Part_3_DeepONetCartesianProd_DiffusionEquation.ipynb** this title does not match the content which shows FNO. It looks like deprecated file to be deleted.
> 
> **Part_4_FNO_1DAllenCahn.ipynb** The role of convolution layers should be explained.



FNO examples:
**Part_4_FNO_1DAllenCahn.ipynb** -> **Part_3_FNO_1DAllenCahn.ipynb**
**Part_4_FNO_DarcyFlow.ipynb** ->  **Part_4_FNO_DarcyFlow.ipynb** - explain the embeddings
**Part_6_FNO_DiffusionEquation.ipynb** -> **Part_5_FNO_DiffusionEquation.ipynb** explain the diff with **Part_4_FNO_DarcyFlow.ipynb**

Parv621 committed:
File renames as of 12/17 meeting

Parv621 committed:
mardown changes part1 - as discussed in 12/17 meeting

Parv621 committed:
Changes to .py scripts to match notebooks

Parv621 committed:
README updated for Colab notebooks

Parv621 committed:
markdown changes part2 - doscvussed as of 12/17 meeting

Parv621 committed:
Remove unsolved notebooks

Parv621 committed:
DeepXDE poisson1D example

Parv621 committed:
Updates PI DeepONet Poisson

Parv621 committed:
Attempts with Poisson problem - wont converge without LBFGS

Parv621 committed:
Trying to make Advection example work

Parv621 committed:
Working DeepXDE PIDeepONet example for 1D Advection

Parv621 committed:
LBFGS working in Poisson PIDeepONet

Parv621 committed:
file name cleanup

Parv621 committed:
Final cleanups completed, markdown added for PIDeepONet

Parv621 committed:
Python script for PIDeepONet Advection notebook

Parv621 committed:
Add dev loader, enrich trunk space, train till 2000 epochs - notebook works well

Parv621 committed:
Updated script for PIDeepONet example

Parv621 committed:
Update README with (false) links

Parv621 committed:
Update wrappers for all NO models from neuralop

Parv621 committed:
Working SFNO example for Spherical Shallow Water equations, from neuralop library

Parv621 committed:
Notebook import changes

Parv621 committed:
docstring cleanups

Parv621 commented:
Updated the PR with SFNO examples as well ([SFNO on the spherical Shallow Water equations](https://neuraloperator.github.io/dev/auto_examples/models/plot_SFNO_swe.html))

Parv621 committed:
TOML file changes

drgona reviewed (changes requested):
**Examples**
[Part_2_DeepONet_antiderivative_unaligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_2_DeepONet_antiderivative_unaligned.ipynb)
typos in markdown: 
"We use dde.nn.DeepONetCartesianProd to create the DeepONet directly for DeepXDE library" -> this should be 'dde.nn.DeepONet'

[Part_3_FNO_1DAllenCahn](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_3_FNO_1DAllenCahn.ipynb)

typo in title: Operator Learing -> Operator Learning

typos in markdown: Problem Description - Allen-Cahn equation
- "$(u^3 - u)$ is the derivative of the double-well potential..." should be it own bullet point 
- also lets define the diffusion term explicitly 
- In Allen–Cahn the coefficient \epsilon^2 is typically multiplying the diffusion term, alternatively we can write 1/\epsilon^2 in front of the potential

specify the data: domain, boundary conditions, what input and output represent

[Part_4_FNO_DarcyFlow](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_4_FNO_DarcyFlow.ipynb)

- The problem description can be confusing. Your Problem Description introduces “1D Darcy law” and then immediately uses the 2D elliptic PDE form. So it is confusing whether the example is 1D or 2D. We need to clarify.
- specify the data: domain, boundary conditions, what input and output represent

[Part_5_FNO_DiffusionEquation](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_5_FNO_DiffusionEquation.ipynb)

- add relative path hyperlink to Darcy Flow example
- in the introduction markdown you use u, and in the loss definition you use y for outputs. Consolidate, double check the same in other examples

[Part_6_PIDeepONet_Advection_aligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb)

- This equation could benefit from mass conservation loss. 
$\displaystyle \int u(x,t)\,dx = \text{const.}$
this could be implemented by: 
1. constraining mass to equal the initial mass
2. penalize time derivative of mass $\displaystyle \int_\Omega \partial_t u(x,t)\,dx \approx 0$
3. Normalize mass via the relative mass conservation loss as
\[
L
= \frac{1}{N_t}\sum_{j=1}^{N_t}
\left(
\frac{M_\theta(t_j) - M_0}{|M_0| + \varepsilon}
\right)^2.
\]
$\varepsilon > 0$ is a small constant added for numerical stability

- we coul plot phase error: shift error vs time
- Advection without specifying BCs is ill-posed.  Why are we not imposing BC in the loss?

[Part_7_SFNO_ShallowWater](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_7_SFNO_ShallowWater.ipynb)

- Add didactical markdown to explain the SFNO architectyre
- Add reference to the [paper](https://arxiv.org/abs/2306.03838)
- explain the PDE problem in sufficient detail, add equations similarly as in other examples
- define operator-learning task


**Core**
Core changes in trainer.py, problem.py:
Add 1–2 lightweight unit tests that verify:
- training loop accepts dev_loader/validation dataset 
- callback doesn’t crash when dev data is absent/present
- Ensure default behavior is unchanged when no validation dataset is provided.

Parv621 committed:
Markdown changes as requested in PR265

Parv621 committed:
Training at 16x16 -> Test at 32x32

Parv621 committed:
SFNO markdowns

Parv621 committed:
Darcy Flow - fastest loss approach as default, markdwon fixes

Parv621 committed:
Added pytest test_trainer_optional_dev_data, helper function RecordingCallback. Formatted files

Parv621 committed:
Add pytest test_problem_predict

Parv621 commented:
Everything apart from changes suggested in [Part_6_PIDeepONet_Advection_aligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb) is complete. 

Good point regarding the ill-posedness. The BCs are actually handled implicitly via the data distribution. By using the ExpSineSquared kernel, all generated initial conditions are periodic. This effectively imposes Periodic BCs, resolving the ill-posedness.

I will add a small markdown to the notebook

drgona commented:
> * n could benefit from mass conservation loss.

no need to apply the suggested changes to [Part_6_PIDeepONet_Advection_aligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb) - small markdown will be nice

lets focus on the pytests so we can deploy this PR today :)

Parv621 committed:
Updated PIDeepOnet markdown, epoch counts

Parv621 committed:
Added tests for neuralop library wrapper class (_StripMetadataMixin) in operators.py

Parv621 commented:
> > * n could benefit from mass conservation loss.
> 
> no need to apply the suggested changes to [Part_6_PIDeepONet_Advection_aligned](https://github.com/SOLARIS-JHU/neuromancer_NO/blob/feature/FNO/examples/neural_operators/Part_6_PIDeepONet_Advection_aligned.ipynb) - small markdown will be nice
> 
> lets focus on the pytests so we can deploy this PR today :)

Added the following pytests:
- `test_trainer_optional_dev_data` in test_trainer.py to check for running with and without dev data
- `test_problem_predict` in test_problem.py to check the `predict` method in problem.py (grad off, check input/ouput keys)
- `test_strip_state` and `test_strip_load` in test_operators.py, to check the methods in the neualoperator library wrapper `class _StripMetadataMixin` in operators.py

I will also update the main description accordingly

Parv621 commented:
Pytests ran:

test_problem.py, test_operators.py
<img width="1242" height="703" alt="image" src="https://github.com/user-attachments/assets/49d4c159-4d6d-4afe-bfe3-e85cd7c9d0ec" />

test_trainer.py (reletively time consuming)
<img width="1254" height="1132" alt="image" src="https://github.com/user-attachments/assets/06f0946d-36c5-4e43-858b-d7eddd7c3c91" />

drgona committed:
Decrease simulation size and training epochs in test trainer

Reduced the number of simulation data points and epochs for testing trainer to reduce runtime

drgona reviewed (approved):
tests passing on my end, PR is ready to bedeployed

Parv621 committed:
DeepXDE wrapper tests

Parv621 committed:
DeepONetCartesianProd tests
